### PR TITLE
Represent cell States

### LINKED ISSUE #1: Represent Cell State
Model the state a cell in an hexagonal akari has.
Include the possibility of cells being crossed out.

### PR DESCRIPTION
**Pull request by jimgerth**
Add `cell` module with types to represent cell state.

Fixes #1.

### CODE REVIEW

File: src/cell.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    /// A free Rokkakari game cell.
+    /// Its state is represented by [`free::State`](free::State).
+    Free(free::State),
+
+    /// A wall Rokkakari game cell.
+    /// Its state is represented by [`wall::State`](wall::State).
+    Wall(wall::State),
+}
+
+pub mod free {
```
Comment (jimgerth) [inline src/cell.rs]: Consider moving this submodule to its own file `src/cell/free.rs`, especially when this module grows more.
Replace this `mod` block with `pub mod free;` then.

File: src/cell.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        marking: Option<Marking>,
+    }
+
+    pub enum Marking {
+        Light,
+        Cross,
+    }
+}
+
+pub mod wall {
```
Comment (jimgerth) [inline src/cell.rs]: Same goes for this submodule.

File: src/cell.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+//!
+//! ## Constaints
+//!
+//! There are two types of wall cells: *constrained* and *unconstrained* ones.
+//! *Constrained* wall cells contain a number from `0` to `3`.
+
+/// The `Cell` type. See [the module level documentation for more](self).
+pub enum Cell {
+    /// A free Rokkakari game cell.
+    /// Its state is represented by [`free::State`](free::State).
```
Comment (jimgerth) [inline src/cell.rs]: Remove this doc comment line. The `free::State` gets linked automatically.

File: src/cell.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    Free(free::State),
+
+    /// A wall Rokkakari game cell.
+    /// Its state is represented by [`wall::State`](wall::State).
```
Comment (jimgerth) [inline src/cell.rs]: Same here, remove this line.

**Review by jimgerth — COMMENTED** [review]
Be sure to add more types or fields to existing types to really represent all possible cell states. Also do not forget to document.

**Comment by jimgerth** [comment]
Rebased on `master`.

**Review by jimgerth — COMMENTED** [review]
Move the changes in `sec/cell/mod.rs` to the previous commit.

File: src/cell/mod.rs [hunk 5]
```diff
@@ -36,3 +36,16 @@
 //!
 //! There are two types of wall cells: *constrained* and *unconstrained* ones.
 //! *Constrained* wall cells contain a number from `0` to `3`.
+
+// Exported Types
+pub use self::cell::Cell;
```
Comment (jimgerth) [inline src/cell/mod.rs:58]: Move each of these `use` statements to the commits that intoduce the corresponding module.

File: src/cell/mod.rs [hunk 6]
```diff
@@ -36,3 +36,16 @@
...
+pub use self::free::{Marking, State as FreeState};
+pub use self::side::Side;
+pub use self::wall::{Constraint, State as WallState};
+
+// Internal Modules
+#[allow(clippy::module_inception)]
+mod cell;
```
Comment (jimgerth) [inline src/cell/mod.rs:65]: Same with these `mod` statements.

File: src/lib.rs [hunk 7]
```diff
@@ -42,3 +42,5 @@
 //! [rokkakari_puzzle]: https://www.janko.at/Raetsel/Varianten/020.a.htm
 //! [rokkakari_demo]: https://jimgerth.herokuapp.com/akari
 //! [github_repo]: https://github.com/JimGerth/rokkakari
+
+pub mod cell;
```
Comment (jimgerth) [inline src/lib.rs:46]: Add to the commit that originally introduced `src/cell/mod.rs`.

**Comment by JimGerth** [comment]
Remove ticks (`) from commit messages. GitHub does not seem to support mark down in the messages.

File: src/cell/free.rs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    /// A *lamp* placed in a free cell lights up other free cells in every
+    /// direction.
+    Lamp,
+
+    /// A *cross* placed in a free cell is used to signify that there
+    /// cannot be a lamp in that cell.
+    Cross,
+}
+
+struct Illumination {
```
Comment (jimgerth) [inline src/cell/free.rs]: Add documentation!

File: src/cell/mod.rs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+//!
+//! - ...are the ones typically painted *black* in regular [Akari].
+//! - ... do not need to be *- or rather cannot be -* lit up to solve the puzzle.
+//! - ...contain *neither* light sources *nor* crosses.
+//! - ...block off the light coming from a light source in a free cell for other free cells.
+//! - ...***can*** constrain the number of light sources *directly* adjacent to them.
+//!
+//! [akari]: https://en.wikipedia.org/wiki/Light_Up_(puzzle)
+//!
+//! ### Constraints
```
Comment (jimgerth) [inline src/cell/mod.rs:40]: Add further documentation.

File: src/cell/free.rs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    a: bool,
+    b: bool,
+    c: bool,
+    d: bool,
+    e: bool,
+    f: bool,
+}
```
Comment (jimgerth) [inline src/cell/free.rs]: Maybe add a function like `Illuminated_from(side: Side) -> bool`. Also think about if this should rather be implemented on the `cell::Cell` type and how exactly. Should it just return false if the cell is a `Wall` cell? Should there still be a function here, that is private though but gets used by the public function on `cell::Cell`? Should this `FreeState` even be public?

File: src/cell/side.rs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    C,
+    #[allow(clippy::missing_docs_in_private_items)]
+    D,
+    #[allow(clippy::missing_docs_in_private_items)]
+    E,
+    #[allow(clippy::missing_docs_in_private_items)]
+    F,
+}
+
+impl Side {
```
Comment (jimgerth) [inline src/cell/side.rs]: Remove `impl` block.

File: src/cell/free.rs (hunk 10 above)
Comment (jimgerth) [reply src/cell/free.rs]: This not moved to #10.

**Pull request merged**